cyfra-e2e-test/src/test/scala/io/computenode/cyfra/e2e/Fs2Tests.scala

```diff
@@ -0,0 +1,215 @@
+package io.computenode.cyfra.e2e
+
+import io.computenode.cyfra.core.aalegacy.*, mem.*, GMem.fRGBA
+import io.computenode.cyfra.core.layout.Layout
+import io.computenode.cyfra.core.GBufferRegion, GBufferRegion.BufferFromRam
+import io.computenode.cyfra.core.{GExecution, GProgram}
+import io.computenode.cyfra.dsl.{*, given}, gio.GIO, algebra.VectorAlgebra, buffer.GBuffer
+import struct.GStruct, GStruct.Empty, Empty.given
+
+import fs2.{io as fs2io, *}, fs2io.file.{Path, Files}
+import cats.effect.IO
+
+import java.nio.{ByteBuffer, file}, file.{Path as nioPath, Files as nioFiles}
+import org.lwjgl.BufferUtils
+import izumi.reflect.Tag
+import scala.reflect.ClassTag
+import io.computenode.cyfra.spirv.SpirvTypes.typeStride
+
+object Fs2:
+  trait Bridge[CyfraType <: Value: FromExpr: Tag, ScalaType: ClassTag]:
+    def toByteBuffer(inBuf: ByteBuffer, chunk: Chunk[ScalaType]): ByteBuffer
+    def fromByteBuffer(outBuf: ByteBuffer, arr: Array[ScalaType]): Array[ScalaType]
+
+  given Bridge[Int32, Int]:
+    def toByteBuffer(inBuf: ByteBuffer, chunk: Chunk[Int]): ByteBuffer =
+      inBuf.asIntBuffer().put(chunk.toArray[Int]).flip()
+      inBuf
+    def fromByteBuffer(outBuf: ByteBuffer, arr: Array[Int]): Array[Int] =
+      outBuf.asIntBuffer().get(arr).flip
+      arr
+
+  given Bridge[Float32, Float]:
+    def toByteBuffer(inBuf: ByteBuffer, chunk: Chunk[Float]): ByteBuffer =
+      inBuf.asFloatBuffer().put(chunk.toArray[Float]).flip()
+      inBuf
+    def fromByteBuffer(outBuf: ByteBuffer, arr: Array[Float]): Array[Float] =
+      outBuf.asFloatBuffer().get(arr).flip()
+      arr
+
+  given Bridge[Vec4[Float32], fRGBA]:
+    def toByteBuffer(inBuf: ByteBuffer, chunk: Chunk[fRGBA]): ByteBuffer =
+      val vecs = chunk.toArray[fRGBA]
+      val size = vecs.length
+      vecs.foreach:
+        case (x, y, z, a) =>
+          inBuf.putFloat(x)
+          inBuf.putFloat(y)
+          inBuf.putFloat(z)
+          inBuf.putFloat(a)
+      inBuf.flip()
+      inBuf
+
+    def fromByteBuffer(outBuf: ByteBuffer, arr: Array[fRGBA]): Array[fRGBA] =
+      val res = outBuf.asFloatBuffer()
+      for i <- 0 until arr.size do arr(i) = (res.get(), res.get(), res.get(), res.get())
+      outBuf.flip()
+      arr
+
+  extension (buf: ByteBuffer)
+    def put[C <: Value: FromExpr: Tag, S: ClassTag](chunk: Chunk[S])(using bridge: Bridge[C, S]): ByteBuffer =
```

Comment:
I'd recommend we just use `fromByteBuffer`, `toByteBuffer` directly from Bridge. Those extension methods can be slightly misleading, as they appear to come from the original java API and add an extra step to follow when reading the code.